File: 599. Minimum Index Sum of Two Lists/solution.py
class Solution:
    def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
        map1 = {x: i for i, x in enumerate(list1)}
        rsum, rlist = float('inf'), []
        for i, rest in enumerate(list2):
            if i > rsum:
                break
            if rest in map1:
                if i + map1[rest] < rsum:
                    rlist = [rest]
                    rsum = i + map1[rest]
                elif i + map1[rest] == rsum:
                    rlist += [rest]
        return rlist

File: 599. Minimum Index Sum of Two Lists/solution.go
func findRestaurant(list1 []string, list2 []string) []string {
	AndyMap := make(map[string]int)
	for idx, item := range list1 {
		AndyMap[item] = idx
	}

	minIndex := len(list1) + len(list2)
	ret := make([]string, 0)
	for idx, item := range list2 {
		if value, ok := AndyMap[item]; ok {
			curIndex := idx + value

			if curIndex < minIndex {
				ret = []string{item}
				minIndex = curIndex
			} else if curIndex == minIndex {
				ret = append(ret, item)
			}
		}
	}
	return ret
}